### PR TITLE
#0: SDPA Const Max Optimization For Prefill (can work for decode as well)

### PR DESCRIPTION
**Pull request by alingTT**
### Problem description
Currently, in order to maintain the correct online softmax output we have to multiply the intermediate outputs cb_out_acc_im with the cb_exp_max_diff=exp(prev_max - cur_max) for every k chunk we have. From my understanding this ultimately cancels out the previous max factor multiplied and replaces the factor with the most updated max factor. And we have to do this for both the overall exp sum cb_prev_sum and the accumulate outputs. But I believe there is a better way to maintain correctness in the output without actually doing these steps since in our flash attention report we mention exp is one of the largest ops taking up 25% of runtime….

At first I think these steps are needed was because on a hardware level doing exp(x - max(x)) @ V is significantly different from performing (exp(x) @ V) * e^(-max(x)) because of numerical overflow such that we would get extremely large values close to +inf prior to doing the matmul and would give incorrect values, so we must subtract by the updated or current max first so that the exp exponent must always maintain something stable before doing the matmul 

But if that is the sole reason, I believe it is possible to instead of doing this multiplication with cb_exp_max_diff for every chunk, we can safely omit this step and replace the QK -= cb_cur_max and QK = exp(QK) step with exp(x-44.25) (Essentially wherever exp(x-curr_max), we replace curr_max with a constant with a restriction on x to prevent overflow) while still keeping track of curr_max, the reason for 44.25 (half of 88.5) is such that for the operating range of x (from -88.5 to 88.5) that we care about, it still ensures stability, then we only do one multiplication at the very end of the compute kernel with exp(44.25- curr_max) (essentially a one time cb_exp_max_diff ). Mathematically the outputs and the softmax sum are still being scaled by the curr_max and it should not change the distribution of softmax. So it should ultimately be equivalent to the original implementation while saving the cb_prev_sum *= cb_exp_max_diff step and the cb_out_accumulate_im *= cb_exp_max_diff steps so both these steps are only done once, at least for fast-approx mode...

### What's changed
- Instead of subtracting by `curr_max` we subtract by 44.25, but we still keep track of curr_max
- we omit the exp(prev_max - curr_max) calculation and we perform exp(44.25 - curr_max) and this is multiplied at the end of the kernel to the output logits and the sum.
- Upsides:
   - up to 10-11% savings in kernel duration depending on the k chunk sizes
- Downsides:
   - there is potentially a precision issue where we are clipping a certain range of logits, but this can be further investigated as to how we can mitigate this if we have more information about the distribution of the logits prior to exp

Some perf numbers reported here in this [sheet](https://docs.google.com/spreadsheets/d/1XPAlkTKcP68SAhKQ7tcT8b6KkS2m3rTP3IMa275QmIE/edit?gid=1374507100#gid=1374507100)

### CODE REVIEW

**Comment by alingTT** [comment]
closing since this optimization is an approximation and does not necessarily yield consistently good PCC
